process/data_processor.py

```diff
@@ -0,0 +1,155 @@
+    then transforming the testing set
+    """
+
+    def __init__(self):
+        self.mean_vec = None
+        self.idf_vec = None
+        self.events = None
+        self.term_weighting = None
+        self.normalization = None
+
+    def fit_transform(self, X_seq, term_weighting=None, normalization=None):
+        """
+        Fit and transform the training set
+        X_Seq: ndarray,  log sequences matrix
+        term_weighting: None or `tf-idf`
+        normalization: None - not implemented yet
+        """
+        self.term_weighting = term_weighting
+        self.normalization = normalization
+
+        # Convert into bag of words
+        X_counts = []
+        for i in range(X_seq.shape[0]):
+            event_counts = Counter(X_seq[i])
+            X_counts.append(event_counts)
+        X_df = pd.DataFrame(X_counts)
+        X_df = X_df.fillna(0)
+        self.events = X_df.columns
+        X = X_df.values
+
+        num_instance, num_event = X.shape
+        # applies tf-idf if pararmeter
+        if self.term_weighting == "tf-idf":
+            df_vec = np.sum(X > 0, axis=0)
+            self.idf_vec = np.log(num_instance / (df_vec + 1e-8))
+            idf_matrix = X * np.tile(self.idf_vec, (num_instance, 1))
+            X = idf_matrix
+
+        # normalization if parameter
+        if self.normalization == "zero-mean":
+            mean_vec = X.mean(axis=0)
+            self.mean_vec = mean_vec.reshape(1, num_event)
+            X = X - np.tile(self.mean_vec, (num_instance, 1))
+        elif self.normalization == "sigmoid":
+            X[X != 0] = expit(X[X != 0])  # expit is logistic sigmoid for ndarrays
+
+        X_new = X
+        print("Train data shape: {}-by-{}\n".format(X_new.shape[0], X_new.shape[1]))
+        return X_new
+
+    def transform(self, X_seq):
+        """
+        transforms x test
+        X_seq: log sequence data
+        """
+
+        # converts into bag of words
+        X_counts = []
+        for i in range(X_seq.shape[0]):
+            event_counts = Counter(X_seq[i])
+            X_counts.append(event_counts)
+        X_df = pd.DataFrame(X_counts)
+        X_df = X_df.fillna(0)
+        empty_events = set(self.events) - set(X_df.columns)
+        for event in empty_events:
+            X_df[event] = [0] * len(X_df)
+        X = X_df[self.events].values
+
+        # applies tf-idf if pararmeter
+        num_instance, _ = X.shape
+        if self.term_weighting == "tf-idf":
+            idf_matrix = X * np.tile(self.idf_vec, (num_instance, 1))
+            X = idf_matrix
+
+        # applies normalization if parameter
+        if self.normalization == "zero-mean":
+            X = X - np.tile(self.mean_vec, (num_instance, 1))
+        elif self.normalization == "sigmoid":
+            X[X != 0] = expit(X[X != 0])
+        X_new = X
+        print("Test data shape: {}-by-{}\n".format(X_new.shape[0], X_new.shape[1]))
+        return X_new
+
+
+if __name__ == "__main__":
```

Comment:
I think this section should be pulled into another script that calls the above class/funcs. If we're running this on another data type, we'd want to just call the functions.

Additionally, I think this project is supposed to be in notebook format, so it may make sense to abstract this into a .ipynb file (similar to the parse folder) such that there's commentary in the ipynb file which may include any additional massaging required to get it exactly ready for the NN and also including a bit of data exploration (this is the size of the data, with x anomalies in the test/train sets etc etc) which we'll need to present as part of the ppt exercise anyways.

Reply:
ill split this file and ill make a walk through notebook